[FLINK-29472] Add first version of release scripts (#1)
apache/flink-connector-shared-utils: README.md

This is a collection of release utils for [Apache Flink](https://flink.apache.org/) connectors.

# Integration

The scripts assume that they are integrated into a connector repo as a submodule into the connector repo

> under `tools/releasing/<some directory>`.

Author: ```suggestion
under `tools/releasing/shared`.
```

# Usage

Some scripts rely on environment variables to be set.  
These are checked at the start of each script.  
Any instance of `${some_variable}` in this document refers to an environment variable that is used by the respective
script.

## check_environment.sh

Runs some pre-release checks for the current environment, for example that all required programs are available.  
This should be run once at the start of the release process.

## publish_snapshot_branch.sh

Creates (and pushes!) a new snapshot branch for the current commit.  
The branch name is automatically determined from the version in the pom.  

> This script should be called when work on a new major/minor version has started.

Reviewer: ```suggestion
This script should be called when work on a new major/minor version of the connector has started.
```

Creates the jars from the current branch and deploys them to [repository.apache.org](https://repository.apache.org).  
The version will be suffixed with `-${FLINK_MINOR_VERSION}` to indicate the supported Flink version.  
If a particular version of a connector supports multiple Flink versions then this script should be called multiple
times.

## publish_git_tag.sh

Creates a release tag for the current branch and pushes it to GitHub.
The tag will be suffixed with `-rc${RC_NUM}`, if `${RC_NUM}` was set.  
This script should only be used _after_ the `-SNAPSHOT` version suffix was removd via `update_branch_version.sh`.

## update_japicmp_configuration.sh

Sets the japicmp reference version in the pom of the current branch to `${NEW_VERSION}`, enables compatibility checks
for `@PublicEvolving` when used on snapshot branches an clears the list of exclusions.  
This should be called after a release on the associated snapshot branch. If it was a minor release it should
additionally be called on the `main` branch.

# Common workflow

1. run `publish_snapshot_branch.sh`
2. do some development work on the created snapshot branch
3. checkout a specific commit to create a release from

> 4. run `check_environment.sh`

Reviewer: I think it makes more sense to have this as the first step and not the 4th. The `publish_snapshots_branch.sh` uses Maven and git, which are checked for availability in `check_environment.sh`.

Author: Basically the issue is that creating a snapshot branch will sometimes not be done by a release manager.
For example, after the ES 3.0 release we have a v3.0 branch and a main branch on 4.0-SNAPSHOT. v3.1 would be created by _someone_ when we want to make some change that requires a new minor version.

So all the other things the check does just aren't required.

5. run `update_branch_version.sh`
6. run `stage_source_release.sh`
7. run `stage_jars.sh` (once for each supported Flink version)
8. run `publish_git_tag.sh` (with `RC_NUM`)
9. vote on release
10. finalize release or cancel and go back to step 2
11. run `publish_git_tag.sh` (without `RC_NUM`)
12. run `update_japicmp_configuration.sh` (on snapshot branch, and maybe `main`)

# Script naming conventions

| Prefix  | Meaning                                                                |
|---------|------------------------------------------------------------------------|
| check   | Verifies conditions without making any changes.                        |
| update  | Applies modifications locally to the current branch.                   |
| stage   | Publishes an artifact to an intermediate location for voting purposes. |

> | publish | Releases an artifact to a user-facing location.                        |

Author: ```suggestion
| release | Publishes an artifact to a user-facing location.                        |
```
It's a bit of a mess with "publishing" having 2 different meanings.
